### DISCUSSION TITLE
[Feature] Child Template

### DISCUSSION THREAD

**Discussion by scottcwilson** [zencart/zencart#6428] [Ideas]
WordPress has this notion of a Child Theme where you only customize the stuff you want to change and get everything else from the regular theme directory, which can be updated with bug fixes, etc.

It would not be hard to do this for Zen Cart.   If approved I'd be happy to add it.  The first use case would be a child theme for Responsive Classic, which could perhaps be provided as a plugin after the release of 1.5.8.

**Comment by drbyte** [comment]
> where you only customize the stuff you want to change and get everything else from the regular theme directory

Must be pointed out:
Responsive Classic is already a "child theme" in this sense, with template_default being the "regular theme directory".

**Comment by drbyte** [comment]
Before writing code, how do you envision it working? There's not enough detail in your first post to respond to "if approved...".

Many questions arise, including this brief incomplete list:

What "kind" of "customizations" are you trying to support? Just CSS? That's easy today.
How many layers of inheritance do you envisioning supporting? Why?
How would you combine javascript files that duplicate each other, incorrectly replacing newer versions with older versions?
How would you combine CSS directives that the child wants to bypass in its parent?
How would you deal with plugins that say "put these files in your template directory" when really they belong in the parent? 
What would you call the "parent" when it's not really the parent because it's a child of template_default?
How would files outside the `/includes/templates` directory be handled, such as modules and language files? (Particularly wishing to avoid multiple nested levels of new override directories.)

**Comment by drbyte** [comment]
More importantly: 
What's the real-world-use-case that drove you to post this in the first place? What were the very specific technical problems that were being encountered? 
Particularly, what files did you need to replace/update that were problematic and thus drove enough frustration to make a post?
The background behind every feature request is way more important than the request itself. I'd rather see those details before seeing anything about implementation ideas for the "how".

**Comment by scottcwilson** [comment]
This would be for code files; css and JS would still have to be replicated.  The spirit of this idea is to allow people to modify just the code files needed. 

The problem it solves is having people copy responsive_classic and then never merge back in the changes. 

>How would files outside the /includes/templates directory be handled, such as modules and >language files? (Particularly wishing to avoid multiple nested levels of new override directories.)

The child template would have overrides directories (named after it) as needed.  It just wouldn't require *all* the responsive_classic files to be copied, as is the case now.

**Comment by barco57** [comment]
The only use case I can come up with off the top of my head is where you might want child templates of "your template" to change the look of the site seasonally, so mostly CSS changes in the children. 
So for most of the year, you just use responsive_classic (or a clone of responsive classic (before dbltoe comes in to remind us of that plugin)), and then you would have children named, responsive_classic-christmas, responsive_classic-easter, etc. Now I know some will say just make multiple clones and edit the clones for those holidays, seems reasonable until you remember the clones bring EVERYTHING, you may not want everything, as you move from default to responsive classic to children you want less in each. Add a plugin to the responsive classic and you now have to add it to every clone,.. if using a children method and the children are lightweight than you possibly won't have to address the plugin in the children. 

I would personally extend this as it is something I have on the back burner that I have been thinking about for years and children would take us partway there. A store owner may want to "rearrange" some stuff for Christmas, change the order of centerboxes on the homepage, increase the number of items in those centerboxes, switch out text on the homepage, etc... So I think it would be brilliant if in the children's template folder there would be files that could override layout config settings. Nothing worse than switching a template and then having to go fiddle with all of those config switches drbyte hates so much and then having to switch them back when the holiday is over and you switch back to the "normal" template

just my two cents and if you take a drink every time you see the word children in this thread it will make you feel better...lol

**Reply by drbyte** [reply]
> A store owner may want to "rearrange" some stuff for Christmas, change the order of centerboxes on the homepage, increase the number of items in those centerboxes, switch out text on the homepage, etc... So I think it would be brilliant if in the children's template folder there would be files that could override layout config settings.

This should be possible using the new `template_settings.php` feature, both in v2.0 and in this Child template feature in v2.1.

**Comment by scottcwilson** [comment]
> Particularly, what files did you need to replace/update that were problematic and thus drove enough frustration to make a post?

I think the approach I am suggesting would make it easier for people to stay up to date.

>Responsive Classic is already a "child theme" in this sense, with template_default being the "regular theme directory".

Agreed - so an alternate approach would be to overrwrite the contents of template_default with what's in responsive_classic, so the latter would become the "real" base.

**Comment by barco57** [comment]
> so an alternate approach would be to overrwrite the contents of template_default with what's in responsive_classic, so the latter would become the "real" base.

if I am understanding you correctly then ummm, no.

**Comment by scottcwilson** [comment]
Can you elaborate on your reasoning?  I am suggesting this because desktop only templates make no sense anymore.  What's the counterargument?

**Comment by barco57** [comment]
so you are saying responsive_classic files should replace template_default files, basically making resposive_classic the base for all things, unless you override absolutely everything in a template to get away from responsive classic?

Let me lay my bias out on the table. I hate (it is a strong word but I couldn't think of another I would use in polite company) responsive classic, am I forced to use it, sure, and I hold my nose the whole time. Where is my poop emoji when I need it.....

**Comment by drbyte** [comment]
> people copy responsive_classic and then never merge back in the changes

True. There's all kinds of things they never merge back in until something breaks.

> overrwrite the contents of template_default with what's in responsive_classic

I knew that would come up. :(  I actually expected it to be said sooner than this.

> What's the counterargument?

IMO the counter-argument is that responsive_classic uses the wonky MobileDetect stuff to determine mobile mode, instead of being truly responsive. That and other bloat it introduced that I wish we'd done differently.

> that could override layout config settings

I'm toying with ripping out the defaults for those and creating a json field in a template-related table to store template-specific configs for those. It's a lotta work, and a lotta admin ui dev, and a pain to make backward-compatible with existing templates. But it would reduce a lot of the pain of settings being suddenly wrong when switching between templates.
But this is a serious digression from the topic at hand.

**Comment by drbyte** [comment]
I'd like to envision more clearly what overrides a "child template" would have in this context. 

I'd prefer real examples, not contrived/theoretical.

And, while I don't want a rabbit trail, for completeness I should ask: how would it differ from the `/shared` template tier that was experimented with in v160-alpha? I'm pretty sure you're not talking about the same thing, but would like to know the intersection of the two.

**Comment by barco57** [comment]
> wonky MobileDetect stuff to determine mobile mode

yes, mobile detect can be tossed to the curb

> I'm toying with ripping out the defaults for those and creating a json field in a template-related table to store template-specific configs for those. It's a lotta work, and a lotta admin ui dev, and a pain to make backward-compatible with existing templates. But it would reduce a lot of the pain of settings being suddenly wrong when switching between templates.

Yes, please. Draw a line in the sand and don't worry about backward-compatible.

**Comment by scottcwilson** [comment]
> And, while I don't want a rabbit trail, for completeness I should ask: how would it differ from the `/shared` template tier that was experimented with in v160-alpha? I'm pretty sure you're not talking about the same thing, but would like to know the intersection of the two.

Shared was another idea for multiple templates.   Different.

**Comment by scottcwilson** [comment]
> I'd like to envision more clearly what overrides a "child template" would have in this context.

Consider the following workflow: 
- Create template rc_child (child of responsive_classic.  (This creates includes/templates/rc_child/template_info.php)

You then create the following customizations: 
- includes/templates/rc_child/css/stylesheet.css 
- includes/languages/rc_child/english.php 
- includes/languages/english/rc_child/index.php

of course, instead of basing your template on responsive classic, it could be some other template - the principle is the same. 

If there's an update to the template, you drop it in and only need to worry about merging files you have changed.

**Comment by Zen4All** [comment]
> > And, while I don't want a rabbit trail, for completeness I should ask: how would it differ from the `/shared` template tier that was experimented with in v160-alpha? I'm pretty sure you're not talking about the same thing, but would like to know the intersection of the two.
> 
> Shared was another idea for multiple templates. Different.

Liked it, and backported to several 1.5.x projects

**Comment by lankeeyankee** [comment]
> Yes, please. Draw a line in the sand and don't worry about backward-compatible.

This!!

**Comment by mc12345678** [comment]
>> Particularly, what files did you need to replace/update that were problematic and thus drove enough frustration to make a post?

> I think the approach I am suggesting would make it easier for people to stay up to date.

As to the staying up-to-date, I'm not really seeing anything above that improves on that. In fact now there are possibly two files that need to be updated, the parent and the child. I see that it would be easier to update if there was a way to more easily align the template override files with the default/parent files.  Right now this is or can be done mostly folder-by-folder rather than template-to-template. I've found a way around that, but I also see that too often individuals are recommended in the forum to ditch their old template and either start anew or with something newer. Statement is made for various reasons whether legitimate or not, but it partially surprises me that keeping a template up-to-date is being identified as I see it as pretty much the primary concern. I'm glad to see it, but surprised.

I do also get the "seasonal" template, I tried that early on in my working with Zen Cart and thought I had it all figured out, but then too came the "plugins" part of ensuring everything was everywhere it needed to be, worked pretty well for the year used, but haven't done that again since. It *was* a full different template rather than just say a color change. Would be willing to revisit the idea of seasonal templating, preferably I guess with minimal file adjustment.

**Comment by scottcwilson** [comment]
> As to the staying up-to-date, I'm not really seeing anything above that improves on that.

It helps because you only copy the files you change, rather than ALL of them.  So on the next upgrade, if a file you haven't changed is fixed, you get the fix without needing to do any merging or copying.

**Comment by drbyte** [comment]
I'd forgotten to ask this earlier:

Would it make sense to think in terms of "parent" instead of "child"? 
eg: "this template, which is set as Current, names its 'parent' as XYZ template, so lookups go there first before template_default"?

Or is that what you had in mind but just hadn't stated it that way?

A question arises then: to what degree does any version-control become crucial when naming the parent?

**Comment by scottcwilson** [comment]
I used the nomenclature that WordPress uses.  That's all.

**Comment by drbyte** [comment]
> A question arises then: to what degree does any version-control become crucial when naming the parent?

**Comment by scottcwilson** [comment]
Parent is always the latest and greatest.  Child is where merging needs to be done to pull in changes.

**Comment by lat9** [comment]
Here's my current 'take' on the changes required for this feature:
- Need to load child/parent/base for languages, modules, sideboxes and templates
- "Child" template CSS loading should pull parent, then child directories.
- "Child" template `jscript` loading should pull child, then parent directories
- See for Wordpress handling: https://developer.wordpress.org/themes/advanced-topics/child-themes/
- Perhaps "Make Child Template" ala "Clone a Template" adds a "my parent is" variable for use on storefront.

## Catalog Files

The following files make use of `$template_dir`:

- includes/classes/
  - template_func.php
  - ResourceLoaders
    - SideboxFinder.  Needs to search child->parent->base.
- includes/functions/
  - functions_files.php
    - zen_get_file_directory
    - zen_get_module_directory
    - zen_get_module_sidebox_directory
    - zen_get_index_filters_directory
  - html_output.php
    - zen_image_OLD
    - zen_image
- includes/init_includes/init_templates.php
  - Sets DIR_WS_TEMPLATE_PARENT if a child template is in use, also sets `$parent_template_dir` to contain the name of the parent template.
- includes/modules/
  - column_left.php
  - column_right.php
  - column_single.php
  - extra_definitions.php
- includes/templates/{parent-template}/
  - template_info.php.  Contains an indicator of 'support' for children.  When a 'child' is created, contains an indicator of its 'child' status as well as the name of its parent template.
  - common/html_header.php.  Must be knowledgeable of CSS/JS loading "rules" for its children.

## Admin Files

Changes needed to:

- layout_controller.php. 
  - Must be 'aware' of child templates, including child/parent/base sideboxes for enablement.
  - Makes use of change to the SideboxFinder class.
- template_select.php. 
  -  Must be 'aware' of parent/child templates.
  - Offers "Create Child" for children-supporting parent-templates.
  - Indicates that a template
    - Supports creation of children
    - Is a 'child' template.  If so, identifies its parent.

**Comment by lat9** [comment]
The to-be-posted comment below serves as the design documentation for the feature.  It'll be updated on an as-needed basis, with the low-level design changes added once the high-level description of the processing is agreed to.

**Comment by lat9** [comment]
# Zen Cart Child Templates

## Overview

The "Child Template" feature is an *extension* of Zen Cart's ~~current~~ **encapsulated** template structure, introduced in Zen Cart 3.0.0 via [PR #7720](https://github.com/zencart/zencart/pull/7720).  For example, the ~~built-in~~ `responsive_classic_plugin` template can be considered a *child* of `template_default`.

Only **encapsulated** template packages can be parents; such a template is identified via its`manifest.php` file.  Specifically, that file contains a 'template' element that indicates its template-directory name and its parent, e.g.

```php

return [
    'pluginVersion' => 'v1.0.0',
    'pluginName' => 'Responsive Classic Plugin Template',
    'pluginDescription' => 'Responsive classic as a distinct selectable encapsulated template package.',
    'pluginAuthor' => 'Zen Cart Development Team',
    'pluginId' => null,
    'zcVersions' => [],
    'changelog' => '',
    'github_repo' => '',
    'pluginGroups' => [],
    'template' => [
        'key' => 'responsive_classic_plugin',
        'baseTemplate' => 'template_default',
    ],
];
```

The above example shows a first-generation parent template, since its `baseTemplate` is specified as 'template_default'. A child of that template would identify as

```php
return [
    'pluginVersion' => 'v1.0.0',
    'pluginName' => 'Responsive Classic Child Template',
    'pluginDescription' => 'Responsive classic as a distinct selectable encapsulated template package. This is a child of that template package.',
    'pluginAuthor' => 'Zen Cart Development Team',
    'pluginId' => null,
    'zcVersions' => [],
    'changelog' => '',
    'github_repo' => '',
    'pluginGroups' => [],
    'template' => [
        'key' => 'responsive_classic_child',
        'baseTemplate' => 'responsive_classic_plugin',
    ],
];
```



### Storefront Considerations

Various Zen Cart classes and functions, when loading `.php` files for a site's currently-active template, attempt to load a file from the template's override directory and, if not found, fall back to loading from a file's base storefront `includes` directory.

| "Base" Directory                                    | "Template Override" Directory                                |
| --------------------------------------------------- | ------------------------------------------------------------ |
| `index_filters`                                     | `index_filters/{template-name}`                              |
| `languages`                                         | `languages/{template-name}`                                  |
| `languages/english`<sup>1</sup>                     | `languages/english/{template-name}`<sup>1</sup>              |
| `languages/english/extra_definitions`<sup>1</sup>   | `languages/english/extra_definitions/{template-name}`<sup>1</sup> |
| `languages/english/html_includes`<sup>1</sup>       | `languages/english/html_includes/{template-name}`<sup>1</sup> |
| `languages/english/modules/order_total`<sup>1</sup> | `languages/english/modules/order_total/{template-name}`<sup>1</sup> |
| `languages/english/modules/payment`<sup>1</sup>     | `languages/english/modules/payment/{template-name}`<sup>1</sup> |
| `languages/english/modules/shipping`<sup>1</sup>    | `languages/english/modules/shipping/{template-name}`<sup>1</sup> |
| `modules`                                           | `modules/{template-name}`                                    |
| `modules/sideboxes`                                 | `modules/sideboxes/{template-name}`                          |
| `templates/template_default`                        | `templates/{template-name}`                                  |

<sup>1</sup> For multi-language sites, `english` could be an additional language, e.g. `spanish`.

When the site's active template is a 'child', the Zen Cart handling of `.php` template-overrides looks _first_ in the 'child' template's directories, then in the 'parent' template's directories and finally in the Zen Cart 'base' directories.  For example, if looking for the file named `additional_images.php` in the `/modules` directory, the handling will look for and load the file if found in `/modules/child`, otherwise loading the file if found in `/modules/parent` and finally attempting to load that file from the `/modules` directory.

While Zen Cart's current template structure provides the above fall-back for `.php` files, all auto-loaded CSS and JS files for a template must reside in the template's `/css` or `/jscript` sub-directories, respectively. The initial implementation of "child" templates will enforce this rule, while follow-on support will provide a more lenient approach.

#### Changes Required for Full CSS/JS Inheritance

A template that recognizes the full parent/child relationship, i.e. non-_legacy_, makes modifications to its auto-loading of various CSS and JavaScript/jQuery files required by that template.  This template-related change will be accompanied by core updates that provides the underlying support.

1. All files sourced in the 'parent' template's `/css` sub-directory are first loaded from that directory, followed by a load of any child-overrides of the same 'type' from the 'child' template's `/css` sub-directory.  In this manner, the styling overrides in the 'child' template take precedence over those present in its 'parent'.
2. Since JavaScript/jQuery files can contain functions and classes that cannot be duplicated, the changes required for the `/jscript` files' load is a bit more complicated.  The load of all "side-wide" `.js` and `.php` files (as well as any template-specific files) present in the `/jscript` sub-directories must build a combined list of files present in the 'child' and 'parent' `/jscript` sub-directories and
   1. If a file is present in the 'child' template, but not in the 'parent', the 'child' template's file is loaded.
   2. If a file is present in both the 'child' and 'parent', the 'child' template's file is loaded as an override of the parent.
   3. If a file is present in the 'parent' template, but not in the 'child', the 'parent' template's file is loaded.

### Admin Considerations

Two admin tools (**Layout Boxes Controller** and **Template Selection**) provide the customization and choice of the site's active template.  Each will require functional changes to support 'child' templates.

#### Template Selection Changes

Once a 'child' template is installed via the admin's `Modules :: Plugin Manager`, it can be selected as the site's active template just like any other!  The challenge is creating that 'child'!

##### Selected Template Display

The **Template Selection** tool's tabular display will be updated to include a column that identifies the selected template's 'parent' directory, either displaying that directory name for 'child' templates or &mdash; for a 'parent' template.

![image-20240504100609907](https://github.com/zencart/zencart/assets/2685585/871229c6-aff3-4e7c-8722-8fe48adb7560)

##### Template Sidebox Display

The sidebox displayed by the **Template Selection** tool will be modified to:

1. Display as a full-viewport width section.
2. Remove the _View Template Settings_ button from the active template's section and add buttons for each of the available templates.  If a template doesn't include a `/template_settings.php`, the associated button will display but be disabled.
3. Add a _Create Child_ button for each of the available 'parent' templates.  For any 'child' templates, add a button to _Copy_ that template.

![image-20240504100423692](https://github.com/zencart/zencart/assets/2685585/cbe93648-896b-4937-a676-6f09a70af137)

When a *Create Child* button is clicked, a modal pop-up is displayed to gather

1. The 'key' to be used for the child's encapsulated template package.
2. The directory name to be associated with the 'child' template.
3. The text to be displayed for the to-be-created 'child' when viewing in the **Template Selection** tool.

Once the admin confirms the 'child' template creation:

1.  The 'child' template is created as a new encapsulated plugin.
2.  Empty template-override sub-directories for the 'child' are created in the plugin (refer to the table at the beginning of this article).
3.  The parent's `/css` and `/jscript` sub-directory contents are copied to the child's like-named plugin template sub-directories.  
4.  The 'parent' template's **Layout Boxes Controller** settings are copied to those for the 'child'.
5.  Upon completion of the child-template's creation, the admin is redirected to `Modules :: Plugin Manager`, where the child template can be installed.

When a *Copy* button is clicked for a 'child' template, a modal pop-up is displayed to gather

1. The 'key' to be used for the copy's encapsulated template package.
2. The directory name to be associated with the copy of that 'child' template.
3. The text to be displayed for the to-be-created 'child' when viewing in the **Template Selection** tool.

Once the admin confirms the 'child' template copy:

1.  Template-override sub-directories for the 'copy' are created in the site's file-system and all contents of the source 'child' are copied to that 'copy'.
2. The source 'child' template's **Layout Boxes Controller** settings are copied to those for the 'copy'.

**Reply by drbyte** [reply]
Agreed.

**Reply by lat9** [reply]
Thanks, I'll give another week for additional comments and then start cogitating on the low-level design implications.

**Reply by drbyte** [reply]
To clarify: this proposed "child template" feature is conceptually a "site-specific-customizations" feature, applied specifically to templates?

Thus it essentially lets a storeowner "install (ie: 'drop in') a[any compatible] template" and not (ie: probably never) have to customize it directly, but instead put all desired customizations into a "child template" (eg: directory structure), limiting the kinds of things that need "touching", "cloning", "configuring", etc specific to the store.

So, allows customizations to be more easily self-contained, avoids the pitfalls of the "clone a template" approach, and of course makes upgrading less onerous because fewer store-specific files need touching.

Just looking to clarify interpretations of what it means. Feel free to correct me if I've misstated or missed something.

**Comment by lat9** [comment]
> To clarify: this proposed "child template" feature is conceptually a "site-specific-customizations" feature, applied specifically to templates?
> 

Yep.

> Thus it essentially lets a storeowner "install (ie: 'drop in') a[any compatible] template" and not (ie: probably never) have to customize it directly, but instead put all desired customizations into a "child template" (eg: directory structure), limiting the kinds of things that need "touching", "cloning", "configuring", etc specific to the store.
> 

Right, and when the 'parent' template is updated any _non-overridden_ files for the 'child' template just 'activate'.

> So, allows customizations to be more easily self-contained, avoids the pitfalls of the "clone a template" approach, and of course makes upgrading less onerous because fewer store-specific files need touching.
> 

Also correct.  Fewer files to check for updates (only those overridden by the 'child').

> Just looking to clarify interpretations of what it means. Feel free to correct me if I've misstated or missed something.

**Comment by lat9** [comment]
**_Note_**: Functional description updated to enable a 'child' template to be copied to another 'child' and for a 'child' template to be removed.

**Comment by gburton** [comment]
Please take a look at how we implemented templating.
It might give some ideas, or show pitfalls to avoid.
And any ideas/issues you come up with would be helpful for us as well.

--

The basis is that we have a "default" template `/templates/default/` and `core files`, which can be extended by other templates, eg "override" example at `/templates/override/`

To easily see it in action:
1. install Phoenix out of the box
2. Now drop a copy of (eg)
`/includes/modules/content/header/templates/tpl_cm_header_logo.php`
into
`/templates/override/includes/modules/content/header/tpl_cm_header_logo.php`
3. amend that new file in any way you wish (eg by adding text or whatever)
4. reload the site in your browser

Changes made in that new file, should show.  In this way, user can do as they want outside the core files...obviously leading to a scenario where a core update _should_ take not more than a few moments to complete.  Give it a go, it'll give you ideas for sure.

Making a new template is super-simple.  I am currently working on a template that radically changes the look of Phoenix - when I'm done, it'll be a "drop in" folder to `/templates/` then back into `admin > my store` and change the used template from "override" to "name_of_new_template".  We think it's a pretty cool system, would be good to hear your thoughts.

**Reply by scottcwilson** [reply]
Thanks for sharing your ideas, Gary!  You exemplify the best of open source.

**Reply by lat9** [reply]
A link to the (a) download and (b) associated documentation would be welcomed!

**Comment by KitchenWarehouse** [comment]
I wish I had something more technical to add except I think this is a great idea 😁

I use a template which is heavily modified, having the ability to use a child theme of this template would massively help when upgrading, currently im looking at moving from 1.5.7d to 2.0.1 and the template modifications are going to be the death of me!

**Reply by scottcwilson** [reply]
A word of encouragement is always good!

**Comment by lat9** [comment]
First PR in preparation: https://github.com/zencart/zencart/pull/6604.  Note that it's not currently [possible](https://github.com/orgs/community/discussions/2971) to link issues/PRs to discussions and vice versa!

That PR splits out the CSS and JS loading sections of a template's `/common/html_header.php` into separate, reusable files.

**Comment by proseLA** [comment]
little late to the party; but my opinion is that a template should reside completely in its own directory.  similar to an encapsulated plugin.

i believe this is how it is done in WP, and then there is a child theme for modifications of said theme.

having template files spread out throughout the entire ZC directory structure is what makes maintenance/upgrades and the like all the more difficult.

implementing a child theme (great idea) without first addressing the encapsulated theme menu seems to me to be putting the cart ahead of the horse.

if that is what gburton is recommending above in his Phoenix example, then i am all for it.

**Comment by lat9** [comment]
As we get closer to having support for "Encapsulated Templates" (see [Zen Cart PR 7720](https://github.com/zencart/zencart/pull/7720), @zcwilt and I have come to a belief that only Encapsulated Templates can be parents, due to the complexity of the inheritance fall-back.  Further, "Child Templates" will also be encapsulated.

**Reply by proseLA** [reply]
> As we get closer to having support for "Encapsulated Templates" (see [Zen Cart PR 7720](https://github.com/zencart/zencart/pull/7720), @zcwilt and I have come to a belief that only Encapsulated Templates can be parents, due to the complexity of the inheritance fall-back. Further, "Child Templates" will also be encapsulated.

other than what is in the repo, i only see 2 people actively working on templates.  so i do not think that this is a problem, and moving templates to complete encapsulation is a great upgrade, IMO, even without child templates.

**Comment by lat9** [comment]
Note that I've just updated that overview (started in May 2024).  Comments and alternate implementation suggestions welcomed.

**Comment by lat9** [comment]
One of the basic requirements, IMO, for the parent/child template relationship is a means for the children to have **different** configuration settings than the parent.  This feature has begun for the zc300 (current master-branch) release via the use of the `TemplateSettings.php` class, instantiated as a globally-available `$tplSetting`.

While that answers the high-level question "how are these settings accessed", it does not address "how do these template-specific settings get recorded in the first place"?

The current use of the `TemplateSettings` class revolves around the creation and edit of a template's `template_settings.php` file.  While that approach "does the trick", it requires a level of programming skills.  For the parent/child template settings to be widely adopted, the implementation needs to provide a higher-level method to gather and store those settings for the non-programmer store-owner.

Some possibilities for discussion
- Update configuration settings to include an indicator as to whether/not a setting can be overridden on a template-by-template basis.
   - An update to the admin's `Configuration` tool would then detect whether any of the settings for the current configuration group can be overridden and, if so, the tool would provide a dropdown of currently-installed templates to choose the template for which the settings' changes would apply.
- Perhaps the configuration tool's updates would be written to the selected template's `template_settings.php` file?

**Reply by drbyte** [reply]
The `template_select` table has a `template_settings` JSON field for storing any configuration values that need to be loaded into the `$tplSetting` object. 
Handled via `init_templates.php` after the optional `template_settings.php` file is loaded.

**Reply by lat9** [reply]
True, but simply writing the template-specific settings to the `template_settings.php` file doesn't require a major shift in the usage of that database field.  That's because the current implementation of the `template_select` table only stores information about the **active** templates' settings.

**Comment by lat9** [comment]
As I continue this discussion with myself ...

Perhaps an update to the "Template Selection" tool would work for the support for the template-specific settings. Here are my current thoughts:
- The Zen Cart core (and an original parent "Template Package") provides a list of configuration settings that can be overridden.
- The "Template Selection" tool is updated to add an "Edit Template Settings" button. Clicking on that button results in a display similar to the "Configuration" tool, but listing all configuration settings that can be overridden.
  - Values displayed in the template-settings' edit display not only the "base" (i.e. in database) settings, a parent's settings and the selected template's settings &mdash; only the template-specific settings are editable.
  - When the settings are saved, they're saved to the selected template's `template_settings.php` file as an associative array (indexed by the configuration_key) named `$tpl_settings`.

That implementation would enable a GUI to change the template-specific settings, making it easier for a store-owner to make the changes themselves.

**Reply by scottcwilson** [reply]
I was thinking the configuration would be accessed under template settings as well.

**Comment by torvista** [comment]
Is a GUI really necessary for everything?

Considering that the person doing the template fiddling should be capable of editing a text file and rational thought, to deal with css and layout changes...

**Reply by barco57** [reply]
Yes, I believe it is necessary. It will allow for template specific configs that are easy for site owner to change. Example being, I build a template with 5 different header layouts which can be quickly selected from within the admin. 
We are trying to increase the use of Zen Cart, which requires making things easier to the store owner to change configs.
